File: autoit-mediapipe-com/src/binding/tasks/vision/gesture_recognizer.cc
#include "binding/tasks/vision/gesture_recognizer.h"

namespace mediapipe::tasks::autoit::core::utils {
	tasks::vision::gesture_recognizer::GestureRecognizerResultCallback CppConvertToGestureRecognizerResultCallback(
		tasks::vision::gesture_recognizer::GestureRecognizerResultRawCallback c_callback
	) {
		return [c_callback](absl::StatusOr<tasks::vision::gesture_recognizer::GestureRecognizerResult> _status_or, const Image& image, int64_t timestamp) {
			auto status_or = tasks::autoit::vision::gesture_recognizer::ConvertToGestureRecognizerResult(_status_or);

			const auto& status = status_or.status();
			const char* error_msg = status.ok() ? nullptr : status.message().data();
			c_callback(&status_or.value(), static_cast<int>(core::utils::ToMpStatus(status)), error_msg, image, timestamp);
		};
	}

	void CppConvertToGestureRecognizerResultCallback(tasks::vision::gesture_recognizer::GestureRecognizerResultRawCallback c_callback, tasks::vision::gesture_recognizer::GestureRecognizerResultCallback& fn, HRESULT& hr) {
		fn = CppConvertToGestureRecognizerResultCallback(c_callback);
		hr = S_OK;
	}

	void CppConvertToGestureRecognizerResultCallback(VARIANT* in_val, tasks::vision::gesture_recognizer::GestureRecognizerResultCallback& fn, HRESULT& hr) {
		using Callback = tasks::vision::gesture_recognizer::GestureRecognizerResultRawCallback;

		Callback c_callback = nullptr;
		hr = autoit_to(in_val, c_callback);
		if (SUCCEEDED(hr)) {
			fn = CppConvertToGestureRecognizerResultCallback(c_callback);
		}
	}
}

PTR_BRIDGE_IMPL(mediapipe::tasks::vision::gesture_recognizer::GestureRecognizerResultRawCallback);

template<typename _In, typename _Out>
inline const HRESULT autoit_to_callback(VARIANT const* in_val, _Out& result_callback) {
	_In c_callback;
	HRESULT hr = autoit_to(in_val, c_callback);
	if (SUCCEEDED(hr)) {
		result_callback = ::mediapipe::tasks::autoit::core::utils::CppConvertToGestureRecognizerResultCallback(c_callback);
	}
	return hr;
}

const HRESULT autoit_to(VARIANT const* in_val, mediapipe::tasks::vision::gesture_recognizer::GestureRecognizerResultCallback& result_callback) {
	using _In = mediapipe::tasks::vision::gesture_recognizer::GestureRecognizerResultRawCallback;
	using _Out = mediapipe::tasks::vision::gesture_recognizer::GestureRecognizerResultCallback;
	return autoit_to_callback<_In, _Out>(in_val, result_callback);
}

File: src/declarations/tasks/vision/gesture_recognizer/gesture_recognizer.js
const ns = "mediapipe::tasks::vision::gesture_recognizer";
const nsProgId = ns.replaceAll("::", ".");
const fqn = `${ ns }::GestureRecognizer`;
const RunningMode = "tasks::vision::core::RunningMode";

module.exports = ({language}) => {
    const progid = `mediapipe.tasks.${ language }.vision.gesture_recognizer`;
    const nsLanguage = `tasks::${ language }::vision::gesture_recognizer`;
    const utils = `mediapipe::tasks::${ language }::core::utils`;

    return [
        // expose gesture_recognizer properties like in mediapipe python
        [`${ progid }.`, "", ["/Properties"], [
            [`${ ns }::GestureRecognizerOptions`, "GestureRecognizerOptions", "", ["/R", "=this", "/S"]],
            [`${ ns }::GestureRecognizer`, "GestureRecognizer", "", ["/R", "=this", "/S"]],
        ], "", ""],

        [`struct ${ nsProgId }.GestureRecognizerOptions`, "", [`/progid=${ progid }.GestureRecognizerOptions`, "/Simple", "/DC"], [
            ["tasks::core::BaseOptions", "base_options", "", ["/RW"]],
            ["core::RunningMode", "running_mode", `${ RunningMode }::IMAGE`, ["/RW"]],
            ["int", "num_hands", "1", ["/RW"]],
            ["float", "min_hand_detection_confidence", "0.5f", ["/RW"]],
            ["float", "min_hand_presence_confidence", "0.5f", ["/RW"]],
            ["float", "min_tracking_confidence", "0.5f", ["/RW"]],
            ["components::processors::ClassifierOptions", "canned_gestures_classifier_options", "", ["/RW"]],
            ["components::processors::ClassifierOptions", "custom_gestures_classifier_options", "", ["/RW"]],
            ["GestureRecognizerResultCallback", "result_callback", "", [`/WExpr=${ utils }::CppConvertToGestureRecognizerResultCallback($value, $0, $hr)`]],
        ], "", ""],

        [`class ${ nsProgId }.GestureRecognizer`, "", [`/progid=${ progid }.GestureRecognizer`], [], "", ""],

        [`${ nsProgId }.GestureRecognizer.Create`, `std::shared_ptr<${ fqn }>`, ["=create_from_model_path", "/S", `/WrapAs=${ utils }::ReleaseToSharedPtr`], [
            ["std::string", "model_path", "", [`/Expr=std::unique_ptr<GestureRecognizerOptions>(new GestureRecognizerOptions{ .base_options{ .model_asset_path{ $0 } }, .running_mode{ ${ RunningMode }::IMAGE } })`]],
        ], "", ""],

        [`${ nsProgId }.GestureRecognizer.Create`, `std::shared_ptr<${ fqn }>`, ["=create_from_options", "/S", `/WrapAs=${ utils }::ReleaseToSharedPtr`], [
            ["std::shared_ptr<GestureRecognizerOptions>", "options", "", ["/Expr=tasks::autoit::core::utils::ConvertToUniquePtr($0)"]],
        ], "", ""],

        [`${ nsProgId }.GestureRecognizer.Recognize`, `absl::StatusOr<${ nsLanguage }::GestureRecognizerResult>`, ["=recognize", `/WrapAs=${ nsLanguage }::ConvertToGestureRecognizerResult`], [
            ["Image", "image", "", []],
            ["std::optional<core::ImageProcessingOptions>", "image_processing_options", "std::nullopt", []],
        ], "", ""],

        [`${ nsProgId }.GestureRecognizer.RecognizeForVideo`, `absl::StatusOr<${ nsLanguage }::GestureRecognizerResult>`, ["=recognize_for_video", `/WrapAs=${ nsLanguage }::ConvertToGestureRecognizerResult`], [
            ["Image", "image", "", []],
            ["uint64_t", "timestamp_ms", "", []],
            ["std::optional<core::ImageProcessingOptions>", "image_processing_options", "std::nullopt", []],
        ], "", ""],

        [`${ nsProgId }.GestureRecognizer.RecognizeAsync`, "absl::Status", ["=recognize_async"], [
            ["Image", "image", "", []],
            ["uint64_t", "timestamp_ms", "", []],
            ["std::optional<core::ImageProcessingOptions>", "image_processing_options", "std::nullopt", []],
        ], "", ""],

        [`${ nsProgId }.GestureRecognizer.Close`, "absl::Status", ["=close"], [], "", ""],
    ];
};
